pkg/controller/core/workload_controller.go

```diff
@@ -313,18 +313,27 @@ func isDisabledRequeuedByClusterQueueStopped(w *kueue.Workload) bool {
 }
 
 func (r *WorkloadReconciler) reconcileCheckBasedEviction(ctx context.Context, wl *kueue.Workload) (bool, error) {
-	if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadEvicted) || !workload.HasRetryOrRejectedChecks(wl) {
+	if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadEvicted) || !workload.HasRetryChecks(wl) && !workload.HasRejectedChecks(wl) {
 		return false, nil
 	}
 	log := ctrl.LoggerFrom(ctx)
 	log.V(3).Info("Workload is evicted due to admission checks")
-	workload.SetEvictedCondition(wl, kueue.WorkloadEvictedByAdmissionCheck, "At least one admission check is false")
-	err := workload.ApplyAdmissionStatus(ctx, r.client, wl, true)
-	if err == nil {
-		cqName, _ := r.queues.ClusterQueueForWorkload(wl)
-		metrics.ReportEvictedWorkloads(cqName, kueue.WorkloadEvictedByAdmissionCheck)
+	if workload.HasRetryChecks(wl) {
+		workload.SetEvictedCondition(wl, kueue.WorkloadEvictedByAdmissionCheck, "At least one admission check is false")
+		err := workload.ApplyAdmissionStatus(ctx, r.client, wl, true)
+		if err == nil {
+			cqName, _ := r.queues.ClusterQueueForWorkload(wl)
+			metrics.ReportEvictedWorkloads(cqName, kueue.WorkloadEvictedByAdmissionCheck)
+		}
+		return true, client.IgnoreNotFound(err)
+	}
+	// if Workload has Rejected checks deactivate it
+	wl.Spec.Active = ptr.To(false)
+	if err := r.client.Update(ctx, wl); err != nil {
+		return false, err
 	}
-	return true, client.IgnoreNotFound(err)
+	r.recorder.Eventf(wl, corev1.EventTypeWarning, "AdmissionCheckRejected", "Deactivating workload because AdmissionCheck for %v has got Rejected", workload.GetRejectedChecks(wl)[0])
```

Comment:
Can you explain where in code we make sure that `workload.GetRejectedChecks(wl)` is not empty to prevent panics?

Also, do we have an analogous event for deactivating due to PodsReady timeout?

Reply:
[At this point](https://github.com/kubernetes-sigs/kueue/pull/2363/files#diff-60dd240c20adbd6a189d018d1c216c2d296730f446c341d8bf449fa6657964ffR330)
We know a Workload has at least one Rejected AdmissionCheck because of previous if statements. I've added a comment clarifying that.

 >> Also, do we have an analogous event for deactivating due to PodsReady timeout?
 Afaik we don't. I can add it in a follow-up PR

Reply:
Good point, fixed

Reply:
I like the change, as it is now also easy to see that `[0]` is safe.